ENH: Introduce compass-notation for legend

ImportanceOfBeingErnest opened the pull request:
## PR Summary

Legend locations (`loc`) can currently be given as strings (like "upper left") or numbers (like `2`). Both have the drawback that they are hard to remember. Some arguments are exchanged about this in #11174

> Is it “top” or “upper”, “bottom” or “lower”, “Center”, “centre”, or “middle”? I’d have to look it up.

> Additionally, the numbers don't even have a reasonable order:

```
    2  9 1
    6 10 7
    3  8 4
```

Also, there is a recent stackoverflow question ["Legend location, similar to telephone"](https://stackoverflow.com/questions/53060095/legend-location-similar-to-telephone), asking why not order the numbers like on a phone.

A possible solution [was proposed by @efiring](https://github.com/matplotlib/matplotlib/pull/11174#issuecomment-386833679) to use a **compass notation**. Also, the [new inset locator](https://github.com/matplotlib/matplotlib/pull/11730) uses compass notation. 

This PR introduces this compass notation, (e.g. `loc="NW"`) 

```
NW   N   NE
W    C    E
SW   S   SE
```

in *addition* to the existing location strings and codes for `legend`. 

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

timhoffm commented:
I'm not really happy with the two-letter compass notation. It's not readable. `loc='SE'` is obscure while `loc='lower right'` is self-explanatory.

For me, the only reason to introduce this is that we already have that in other parts of the library.

For information, MATLAB uses written-out compass notation:
https://de.mathworks.com/help/matlab/ref/legend.html#bt6ef_q-1-lcn
In itself, that's not too bad. However, I doubt we should not introduce yet another format. Also MATLAB has "*outside" locations, which would be really nice, but that's another topic.

anntzer commented:
I think I have a mild preference for the MATLAB-style(!!) spelled-out strings "northeast" etc (as long as we use it everywhere).  Let's say +0.5 for "northeast" and +0 for "NE" (but only one of them, and has to be consistent with eg. inset locator).

timhoffm commented:
I'm afraid we cannot easily get around the short "NE". It's already been used e.g. for anchors https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.set_anchor.html. Even if we decide to use "northeast" everywhere, we'd have to keep "NE" around for the anchor (at least for a deprecation period, but I tend to not deprecating that).

ImportanceOfBeingErnest commented:
Should I allow for "northeast" *in addition* to "NE" then? 

(Note to self: This still needs a *what's new* entry.)

timhoffm commented:
> Should I allow for "northeast" in addition to "NE" then?

I'm +/-0 on this. It's a more readable convention. OTOH, it's not good to have too many ways to specify things.

timhoffm commented on lib/matplotlib/legend.py:
```diff
@@ -112,23 +112,24 @@ def _update_bbox_to_anchor(self, loc_in_canvas):
+        'upper left'     'NW'            2
+        'lower left'     'SW'            3
+        'lower right'    'SE'            4
+        'right'                          5
```
`right` is outside of the Axes. This is what MATLAB calls "outsideeast". Unfortunately, our terminology is not good here, and we simply did not implement any other outside positions.

ImportanceOfBeingErnest replied:
`"right"` and `"center right"` are identical. They both result in the legend to be placed at the "east" position.

https://github.com/matplotlib/matplotlib/blob/9b01a037fb5a4d147ec592e095c887ebfdf2a196/lib/matplotlib/legend.py#L1086-L1088

I suppose one of them only exists for backwards compatibility reasons.  
If you want to place the legend outside the axes you need to supply a tuple of xy coordinates (`loc=(1.05, 0.7)`) or directly use the `bbox_to_anchor` argument.

timhoffm commented:
> I don't think it makes sense to add more than one way to do it for legends. Pick an "ideal" name to use for legends, and then update anchors to support that new ideal name, optionally deprecating the old one.

That's the conflict here. IMO "southeast" is to be preferred to "SE". But "SE" is already used for anchors (https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.set_anchor.html). Different terminologies for legends and anchors make no sense. That would mean "SE" would need to be depreacated for anchors. But OTOH is it worth it?

anntzer commented:
I wouldn't overly worry about set_anchor, which is likely much more rarely used than legend() (and already not consistent).  If we switch to spelled-out compass directions, I'd just do the deprecation dance on set_anchor.

tacaswell commented:
:+1: in general, I am -0 on supporting both the abbreviations and the full names, and +1 on no spaces if we do full names.  I am not sure which way to go on full names vs abbreviations though.

tacaswell commented:
xref https://github.com/matplotlib/matplotlib/pull/13072

tacaswell commented:
>  in general, I am -0 on supporting both the abbreviations and the full names, and +1 on no spaces if we do full names. I am not sure which way to go on full names vs abbreviations though.

I am still :+1: , but on the phone call @efiring convinced me that we should do both and with spaces is going to be easier to read.

anntzer commented:
Strong preference for spaceless longforms.

tacaswell commented:
Following up, on the call again, 

 - allow short notation (no spaces) all uppercase
 - allow long notation (no spaces) all lowercase

ImportanceOfBeingErnest commented on lib/matplotlib/legend.py:
```diff
@@ -584,10 +552,6 @@ def __init__(self, parent, handles, labels,
 
-        tmp = self._loc_used_default
-        self._set_loc(loc)
-        self._loc_used_default = tmp  # ignore changes done by _set_loc
```
Those lines somehow got lost during rebase. I have no idea what they are used for. Anyone else maybe?

jklymak replied:
ping @timhoffm about this code...

timhoffm replied:
This must stay in, and probably get a more extensive comment.

It's a hack to remember if we're using `loc='best'` as default or if it was explicitly set by the user. It's used to determine whether to issue a warning with `loc='best'` and many plot elements. The warning should only be issued with for the default, not when the user explicitly sets 'best'.

See https://github.com/matplotlib/matplotlib/pull/12455#pullrequestreview-176063630
and the link therein: https://github.com/timhoffm/matplotlib/commit/917b5205645af78ea2a419293d147e97cce21422

ImportanceOfBeingErnest commented:
Updated. 
This now introduces "NW" as well as "northwest" as locations for the legend, as well as for other anchored artists and inset_axes. 
For this I externalized the mapping from the possible input values into `cbook`, such that it can be reused everywhere. Internally, the one- or two character strings are now used to specify the location. (Previously strings got mapped to numbers, than back to those compass location codes.)   
This should preserve all deprecations, although there is an attribute `self._loc_used_default` which got lost when rebasing - and I have absolutely no idea about its purpose. 

I introduced getters and setters for `loc` as well. This is a new feature, which allows to set the location a posteriori, and it also ensures that if people used the private attribute `self._loc` in their code to achieve the same, their code would not fail through this change.

anntzer commented:
loc_used_default came from https://github.com/matplotlib/matplotlib/pull/12455.

ImportanceOfBeingErnest commented:
It seems the use cases from #12455 are still covered. And the tests introduced in #12455 pass. So somehow I wonder if I'm missing something.

anntzer commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -2156,3 +2156,65 @@ def _check_in_list(values, **kwargs):
+_valid_compass = ['NE', 'NW', 'SW', 'SE', 'E', 'W', 'E', 'S', 'N', 'C']
+
+
+def _map_loc_to_compass(loc, allowtuple=False, allowbest=False,
```
it's a private api so it doesn't really matter, but these look like good parameters to make keyword only

ImportanceOfBeingErnest commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -2156,3 +2156,65 @@ def _check_in_list(values, **kwargs):
+        if loc in codes:
+            return codes[loc]
+
+    if allowtuple:
```
In case of a legend `loc` can be a 2-tuple (or any list/array of two numbers). The docstrings states 

> The location can also be a 2-tuple giving the coordinates of the lower-left
    corner of the legend in axes coordinates (in which case *bbox_to_anchor*
    will be ignored).

For any other anchored box, this tuple location is not allowed, so the `allowtuple` flag makes sure that the tuple only makes it into the output in case of a legend.

ImportanceOfBeingErnest commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -2156,3 +2156,65 @@ def _check_in_list(values, **kwargs):
+    if loc in _valid_compass:
+        return loc
+
+    if isinstance(loc, str) or isinstance(loc, int):
```
I think if someone were to write e.g. `legend(loc={1:2})` (for whatever misguided reason), `loc in codes` would raise an error, while we would like to have our custom error or warning raised at the end instead.

ImportanceOfBeingErnest commented on doc/users/next_whats_new/comapss_notation.rst:
```diff
@@ -0,0 +1,54 @@
+
+Note that those new compass strings *do not* apply to ``table``.
+
+The above mentioned classes also now have a getter/setter for the location.
```
Ok. Done.

timhoffm commented on lib/matplotlib/legend.py:
```diff
@@ -130,29 +131,31 @@ def _update_bbox_to_anchor(self, loc_in_canvas):
     artists.  This option can be quite slow for plots with large amounts of
     data; your plotting speed may benefit from providing a specific location.
 
+    For back-compatibility, ``'center right'`` (but no other location) can also
```
Simpler:
~~~
For backward-compatibility, ``'right'`` is equivalent to  ``'center right'``;
and there are numeric codes for all locations as well. 
~~~

timhoffm commented on lib/matplotlib/legend.py:
```diff
@@ -130,29 +131,31 @@ def _update_bbox_to_anchor(self, loc_in_canvas):
 
+    For back-compatibility, ``'center right'`` (but no other location) can also
+    be spelled ``'right'``, and each "string" locations can also be given as a
+    numeric value. Possible (case-sensitive) strings and codes are:
```
New paragraph for "Possible ...". Backward-compatibility should be a paragraph of its own.

timhoffm commented on lib/matplotlib/legend.py:
```diff
@@ -621,12 +591,27 @@ def _get_loc(self):
         "Helper function to locate the legend."
 
-        if self._loc == 0:  # "best".
+        if self._loc == 'best':  # "best".
```
```suggestion
        if self._loc == 'best':
```

timhoffm commented on lib/matplotlib/legend.py:
```diff
@@ -1095,26 +1080,12 @@ def _get_anchored_bbox(self, loc, bbox, parentbbox, renderer):
-                        C: "C"}
 
-        c = anchor_coefs[loc]
+        assert loc in _valid_compass  # as this is called only internally
```
Docstring of this internal method needs update on the changed value type of `loc`.

timhoffm commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -2156,3 +2156,64 @@ def _check_in_list(values, **kwargs):
+
+
+def _map_loc_to_compass(loc, allowtuple=False, allowbest=False,
+                        fallback='NE', warnonly=True):
```
Not too important since it's a private API, but I would set the default `warnonly=False`. We actually want to fail for invalid input, and do only have `warnonly` to support a deprecation for default fallback.

With `warnonly=False`, we can already leave out the parameter in cases which are not affected by the deprecation and thus have simpler code there.

timhoffm commented on lib/matplotlib/rcsetup.py:
```diff
@@ -601,16 +601,16 @@ def validate_markevery(s):
+     'center right', 'E', 'east',
+     'lower center', 'S', 'south',
+     'upper center', 'N', 'north',
+     'center', 'C'], ignorecase=True)
```
Add comment that `ignorecase=False` after a deprecation period (i.e. in 3.3).

ImportanceOfBeingErnest replied:
This may be an important point. This PR does not actually deprecate the use of other casing *for rcParams*. If this is desired I would need to change this now.

timhoffm replied:
You will get a deprecation notice when the user for a rcParam value 'Upper Center' when that value is applied, won't you? So it's effectively deprecated. It's only at the end of the deprecation period that the rcParam checker can get more strict and check the case.

ImportanceOfBeingErnest replied:
No, actually, you won't get any warning for rcParameters here because with `ignorecase=True`, they are normalized to the `valid` parameters. 

```
from matplotlib.rcsetup import ValidateInStrings

v = ValidateInStrings("cattle", ["cow", "SHEEP"], ignorecase=True)

print(v("Cow"))     # returns 'cow', no warning
print(v("sheep"))   # returns 'SHEEP', no warning
print(v("Hamster")) # ValueError: Unrecognized cattle string 'hamster'
```

I'm currently thinking about directly using the newly introduced `_map_loc_to_compass` directly for rc validation.

timhoffm replied:
Sounds reasonable.

ImportanceOfBeingErnest replied:
As of now rcParam "legend.loc" only accepts strings (numbers or tuples are not allowed). Is there intention to change that, such that those are allowed as rcParams as well? 

I'm asking since with that parameter running through `_map_loc_to_compass`, 

    plt.rcParams["legend.loc"] = 2

would be allowed, but the same from an rc file

    legend.loc : 2

would fail due to missing string to int conversion - which I could add of course if that is desired.

timhoffm commented on lib/matplotlib/tests/test_offsetbox.py:
```diff
@@ -79,25 +79,39 @@ def test_offsetbox_clip_children():
+
+
+@pytest.mark.parametrize('locs', _loc_values)
+def test_offsetbox_loc_codes(locs):
```
Should this become `@check_figures_equal`? Otherwise the drawing of the figures at the end is meaningless.

ImportanceOfBeingErnest replied:
This was intended as smoke test. A legend with an invalid `_loc` might be constructed, but would fail when being drawn. I don't think image comparisson would help here; however if you are concerned about drawing 6 figures taking too much time, I could externalize this part and run it only the necessary 4 times.

timhoffm replied:
No, 6 times is fine IMHO. Can you please add a short comment on the intent as you've described above?

timhoffm commented on lib/matplotlib/legend.py:
```diff
@@ -621,12 +591,27 @@ def _get_loc(self):
+
+    def get_loc(self):
+        """
+        Get the legend location. This will be one of 'best', 'NE', 'NW',
```
Not sure we want to promote the short compass notation as the canonical API (which it effectively is, if we use it as the return value for `get_loc`.

I'm inclined to leave out  `set_loc`/`get_loc` for now as I'm not sure what the canonical API for `get_loc` should be. And IMHO it's not a too important feature, so that not much harm is done if we're taking this slowly.

Has there been a user request for this or have you added it just because it was straight forward to implement?

ImportanceOfBeingErnest replied:
I don't know of a special user request; but users rarely ask for making private attributes public, instead they just use the private ones. As it is now, this PR does not break any user code that assigns to the private attributes, i.e. `legend._loc = 3` still works as expected. But it would break user code that would compare the location (`if legend._loc == 3: ..`). I'd say this is the risk of using private attributes, but we should provide them with a public alternative, which is `get_loc`.

timhoffm replied:
When we create new public API we should be sure how we want that, and I'm not in this case. This needs discussion with other core devs and should not slip in silently as part of a larger PR. I propose to pull this out into a seprate PR and discuss there.

timhoffm replied:
This was discussed in the call today https://paper.dropbox.com/doc/Matplotlib-2019-meeting-agenda--AYtd1qG~KP8b8TappeAo1YrhAg-aAmENlkgepgsMeDZtlsYu#:h2=12679-compass-notation:

We're not yet clear what the right return value for `get_loc` should be (return the actual user input / or normalized to long compass notation / or normalized to short compass notation).

The main part of this PR is very useful and agreed upon. It would be good get this into 3.1. Please extract `set_loc`/`get_loc` to a separate PR targeted at 3.2 so that we have a bit more time to decide how to do them right.

ImportanceOfBeingErnest replied:
Meaning you want me to introduce private `_get_loc`/`_set_loc` for `AnchoredOffsetbox` as well?

jklymak replied:
I'm not following what you are asking here.  This PR introduces `get_loc` and `set_loc`; they are new, right?  So all we are asking is the new API be delayed so we can talk about it some more.

ImportanceOfBeingErnest replied:
OMG! I think the complete concept breaks down, since 

https://github.com/matplotlib/matplotlib/blob/da701ae6c60c92a4df709b66caf010b8f15b8c06/lib/matplotlib/offsetbox.py#L1029

was public all the time. Hence this PR would break code using the value received by `.loc`.

timhoffm replied:
Not thought this through completely: Are you saying this is a fundamental problem?

ImportanceOfBeingErnest replied:
At least it makes the current state of this PR backwards-incompatible. I need think more on that.

ImportanceOfBeingErnest commented:
>This must stay in (and maybe get a more descriptive comment).  
It's a hack to remember if we're using loc='best' as default or if it was explicitly set by the user. It's used to determine whether to issue a warning with loc='best' and many plot elements. The warning should only be issued with for the default, not when the user explicitly sets 'best'.

What you describe is currently done with this PR - without using those lines. This is also confirmed by the fact that the tests introduced in #12455 do pass.

timhoffm commented:
> What you describe is currently done with this PR - without using those lines. This is also confirmed by the fact that the tests introduced in #12455 do pass.

Passing the tests is a good argument :smile:. Not sure I follow how you do it. Is it the difference between `set_loc` and `_set_loc`? If so, this should be documented in the docstring of `_set_loc`.

ImportanceOfBeingErnest commented:
Whether or not the default is used is stored at instantiation via `self._loc_used_default = loc is None`. The only way for a user to change this is by manually using `set_loc`; `set_loc` contains the line `self._loc_used_default = False`. This is really all there is to it. Possibly the previous implementation assumed that a user might interface with the `_set_loc` instead?

timhoffm commented:
Thing was that previously `self._loc_used_default` was set to False in `_set_loc`. You've moved that to `set_loc`, which seems reasonable to me. I consider the loc issue resolved.

timhoffm commented:
flake8:

./lib/matplotlib/cbook/__init__.py:2170:1: W293 blank line contains whitespace
./lib/matplotlib/cbook/__init__.py:2192:1: W293 blank line contains whitespace
./lib/matplotlib/cbook/__init__.py:2198:1: W293 blank line contains whitespace

timhoffm commented:
Labelled "work in progress" due to the unclear state of backward incompatibility.

ImportanceOfBeingErnest commented:
The problem with backwards incompatibility is the following: 

This PR attempts to simplify the positionning code, such that internally, the short compass notation is used throughout. Previously we would convert from strings to numbers ("lower left" --> 3), then from numbers to compass (3 --> "SW"). With this PR the intermediate step is removed. This works nicely for `legend`, because there is no public method for getting or setting the location.  
Now to the problem: `AnchoredOffsetbox` does have a public `loc` attribute. It's easy to route any assignment to this though the new converter. However, reading out this attribute would yield a different value, i.e. `"SW"` instead of `3`. This is an API change. 

The cleanest solution to this would be to 

* start a deprecation, like *"Warning, this attribute will return the location in compass notation in the future"*; but then the question is how to allow people to adapt their code already now? Usually I would go on with *"To prevent this warning and use the compass notation already now, use the `get_loc` method instead."* However, [some concerns](https://github.com/matplotlib/matplotlib/pull/12679#discussion_r262300065) were raised about adding such `get_loc` method. But without such new method, no useful deprecation can be performed. 

Alternatives are of course:
* Ignore the fact that this is an API change; presumably the number of cases where someone want to read out  `AnchoredOffsetbox.loc` is quite small.
* Reintroduce the conversion from string to number, i.e. let `loc` work with the numbers, only convert to short compass notation at the very end. This would render some of the attempted simplifications of this PR useless.

anntzer commented:
One solution for strict backcompat could be to make the compass locations an IntEnum, which subclass int so they behave like ints for all purposes (see the similar thing I did for MouseButton).

ImportanceOfBeingErnest commented:
I don't think I understand the IntEnum proposal. Would that mean to have some `LocClass.NW` object people need to use instead of `"NW"` or `"northwest"`. Wouldn't that 

(a) result in the same problem of making a decision on the "standard" kind of location, i.e. between `LocClass.NW` and `LocClass.northwest`?
(b) Introduce a hysteresis in arguments you pass (i.e. a string `"northwest"`) versus arguments you get out (i.e. a number/enum `LocClass.NW`)? Or should users then pass `LocClass.NW` themselves?

anntzer commented:
Users would remain able to pass in `"NW"` or `"northwest"` (or `Location.NW`, or 2), but when reading out the loc attribute they would get `Location.NW`.  Because `Location.NW` would be an IntEnum, it would compare equal (for == or as a dict key or whatnot) to its current int value (2).
Essentially, `Location.NW` *is* the number 2, just with a nicer repr.

tacaswell commented:
Pushing to 3.3, but have not read through the full discussion.  If people think this has converged and can be merged in the next ~3 days please do so for 3.2.

ImportanceOfBeingErnest commented:
This is correctly labelled "Work-in-progress" and cannot be merged as is. It will need new input from my side followed by a new round of dicussion.

fourpoints commented:
Here's a suggestion including outside axes with compass notation, essentially treating each cell as a coordinate point from the center. Two steps in any direction is outside the axes. (Admittedly, `northnorth` might sound awkward.)

```
   │NNW│ NN│NNE│
───╆━━━┿━━━┿━━━╅───
WNW┃NW │ N │ NE┃ENE
───╂───┼───┼───╂───
 WW┃W  │ C │  E┃EE
───╂───┼───┼───╂───
WSW┃SW │ S │ SE┃ESE
───╄━━━┿━━━┿━━━╃───
   │SSW│ SS│SSE│
```
